classes/CartRule.php

```diff
@@ -1553,12 +1563,47 @@ public static function cleanProductRuleIntegrity($type, $list)
     public static function getCartsRuleByCode($name, $id_lang, $extended = false)
     {
         $sql_base = 'SELECT cr.*, crl.*
-						FROM '._DB_PREFIX_.'cart_rule cr
-						LEFT JOIN '._DB_PREFIX_.'cart_rule_lang crl ON (cr.id_cart_rule = crl.id_cart_rule AND crl.id_lang = '.(int)$id_lang.')';
+						FROM ' . _DB_PREFIX_ . 'cart_rule cr
+						LEFT JOIN ' . _DB_PREFIX_ . 'cart_rule_lang crl ON (cr.id_cart_rule = crl.id_cart_rule AND crl.id_lang = ' . (int) $id_lang . ')';
         if ($extended) {
-            return Db::getInstance()->executeS('('.$sql_base.' WHERE code LIKE \'%'.pSQL($name).'%\') UNION ('.$sql_base.' WHERE name LIKE \'%'.pSQL($name).'%\')');
+            return Db::getInstance()->executeS('(' . $sql_base . ' WHERE code LIKE \'%' . pSQL($name) . '%\') UNION (' . $sql_base . ' WHERE name LIKE \'%' . pSQL($name) . '%\')');
+        } else {
+            return Db::getInstance()->executeS($sql_base . ' WHERE code LIKE \'%' . pSQL($name) . '%\'');
+        }
+    }
+
+    /**
+     * CartRules compare function to use the Product and the rules
+     *
+     * @param array $arrayProduct List Products,
+     * @param array $arrayRules List Rules,
+     * @param string $type type voucher,
+     *
+     * @return int 0 = same
+     *             1 = different
+     */
+    public function cartRulesCompare($arrayProduct, $arrayRules, $type)
+    {
+        //If the two same array, no verification todo.
+        if ($arrayProduct == $arrayRules) {
+            return $arrayProduct;
+        }
+        $return = array();
+        $array = 0;
+        // Attribute id is not important for this filter in the global list
+        // so the ids are replaced by 0
+        if (in_array($type, array('products', 'categories', 'manufacturers', 'suppliers'))) {
+            $productList = explode(':', preg_replace("#\-[0-9]+#", "-0", implode(':', $arrayProduct)));
         } else {
-            return Db::getInstance()->executeS($sql_base.' WHERE code LIKE \'%'.pSQL($name).'%\'');
+            $productList = $arrayProduct;
         }
+
+        foreach ($productList as $product) {
+            if (in_array($product, $arrayRules) && isset($arrayProduct[$array])) {
+                $return[] = $arrayProduct[$array];
+            }
+            $array++;
+        }
```

Comment:
I'm having a hard time understanding the use for your `$array` variable. I think you could simply do it like this:

```php
$return = array();
foreach ($productList as $k => $product) {
    if (in_array($product, $arrayRules)) {
        $return[] = $arrayProduct[$k];
    }
}

return $return;
```